pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/BaseInstanceSelector.java

```diff
@@ -18,69 +18,194 @@
+ *
+ * New segment is defined as segment that is created more than 5 minutes ago.
+ * - For initialization, we look up segment creation time from zookeeper.
+ * - After initialization, we use the system clock when we receive the first update of ideal state for that segment as
+ *   approximation of segment creation time.
+ *
+ * We retire new segment as old when:
+ * - The creation time is more than 5 mins ago
+ * - We receive error state for new segment
+ * - External view for segment converges with ideal state.
+ *
+ * Note that this implementation means:
+ * 1) Inconsistency across requests for new segments (some may be available, some may be not)
+ * 2) When there is no assignment/instance change for long time, some of the new segments that expire with the clock
+ * are still considered as old.
  */
 abstract class BaseInstanceSelector implements InstanceSelector {
-  private static final Logger LOGGER = LoggerFactory.getLogger(BaseInstanceSelector.class);
+  public static class SegmentState {
+    // List of instance for this segment in ideal state.
+    // Mapping from candidate to index in _candidateInstance.
+    private HashMap<String, Boolean> _candidates;
+
+    private long _creationMillis;
+
+    public SegmentState(long creationMillis) {
+      _creationMillis = creationMillis;
+      _candidates = new HashMap<>();
+    }
+
+    public boolean isNew(long nowMillis) {
+      return CommonConstants.Helix.StateModel.isNewSegment(_creationMillis, nowMillis);
+    }
+
+    public void resetCandidates() {
+      _candidates.clear();
+    }
+
+    public void addCandidate(String candidate, boolean online) {
+      _candidates.put(candidate, online);
+    }
+
+    public HashMap<String, Boolean> getCandidates() {
+      return _candidates;
+    }
+  }
 
   // To prevent int overflow, reset the request id once it reaches this value
   private static final long MAX_REQUEST_ID = 1_000_000_000;
 
-  private final String _tableNameWithType;
   private final BrokerMetrics _brokerMetrics;
+  private final String _tableNameWithType;
+  private final ZkHelixPropertyStore<ZNRecord> _propertyStore;
   protected final AdaptiveServerSelector _adaptiveServerSelector;
 
   // These 4 variables are the cached states to help accelerate the change processing
   private Set<String> _enabledInstances;
   private Map<String, List<String>> _segmentToOnlineInstancesMap;
-  private Map<String, List<String>> _segmentToOfflineInstancesMap;
   private Map<String, List<String>> _instanceToSegmentsMap;
+  private Map<String, SegmentState> _newSegmentStates;
 
-  // These 2 variables are needed for instance selection (multi-threaded), so make them volatile
-  private volatile Map<String, List<String>> _segmentToEnabledInstancesMap;
-  private volatile Set<String> _unavailableSegments;
+  // _segmentStateSnapshot is needed for instance selection (multi-threaded), so make them volatile
+  private volatile SegmentStateSnapshot _segmentStateSnapshot;
```

Comment:
Currently, `_segmentStateSnapshot` is getting updated `onInstanceChange()` and `onAssignmentChange()`. This means that we only update `newSegment` list when there's a change in instance or idealstate/externalview.

If there's no update on instance/idealstate/externalview for a long time (valid case), I think that we currently do not move newSegments to oldSegments. Please correct me if I'm wrong.

If this is the case, should we consider adding a background thread that's periodically check the `SegmentStateSnapshot` and promote newSegments to oldSegments after 5min?

Reply:
this will be a followup PR as discussed

Reply:
👍